src/data/dolma_mixtures.py

```diff
@@ -0,0 +1,204 @@
+    else:
+        dataset.push_to_hub(repo_id, split=split)
+
+
+def main(args):
+    in_fn = args.config_dir / args.config_fn
+    assert in_fn.exists(), f"Config file {in_fn} does not exist."
+    with open(in_fn, 'r') as file:
+        config = yaml.safe_load(file)
+    
+    hf_out_path = config["hf_out_path"]
+    target_tokens = config["target_tokens"]
+
+    api = huggingface_hub.HfApi()
+    if api.repo_exists(repo_id=hf_out_path, repo_type="dataset"):
+        # Ask user if they want to overwrite
+        print(f"Repo {hf_out_path} already exists. Do you want to overwrite?")
+        response = input("y/n: ")
+        if response.lower() != "y":
+            print("Exiting.")
+            return
+        print(f"Ok! Deleting {hf_out_path}.")
+        api.delete_repo(repo_id=hf_out_path, repo_type="dataset")
+
+    existing_cts = pd.read_csv(args.dolma_token_ct_fn)
+    # Number of tokens for each source in existing source data
+    src_cts = dict(zip(existing_cts["source"], existing_cts["num_tokens"]))
+
+    dolma_tok_ct = sum(src_cts.values())
+    print(f"Existing dataset has {dolma_tok_ct} tokens.")
+
+    if target_tokens < 1.0:
+        target_frac = target_tokens
+        target_tokens = round(dolma_tok_ct * target_tokens)
+        print(f"Requested {target_frac} of {dolma_tok_ct} existing tokens ~= {target_tokens} tokens.")
+
+    # Apply coefficients to existing data with no adjustments and measure how close we are to target tokens
+    if config["source_coefficient_type"] == "relative_proportion":
+        sample_coefficients = {x["name"]: x["source_coefficient"] for x in config["sources"]}
+        # Estimate how many tokens we'd have without adjusting sample_coefficients
+        unadjusted_target_tokens = sum([v * src_cts[k] for k, v in sample_coefficients.items()])
+        # Adjust sample_coefficients uniformly to ensure desired target tokens
+        adjustment_ratio = target_tokens / unadjusted_target_tokens
+        sample_coefficients = {k: adjustment_ratio * v for k, v in sample_coefficients.items()}
+    else:
+        assert config["source_coefficient_type"] == "target_fraction", f"Source coefficient type {config['source_coefficient_type']} not recognized."
+        
+        # Compute number of tokens needed to hit target
+        src_toks_needed = {
+            source["name"]: target_tokens * source["source_coefficient"] for source in config["sources"]
+        }
+
+        # Coefficients are the ratio of the number of tokens needed to hit target to the number of tokens we have (src_cts)
+        sample_coefficients = {k: v / src_cts[k] for k, v in src_toks_needed.items()}
+
+    print(f"Targeting {target_tokens} tokens with the following sampling fractions by source:")
+    print("\n".join(f"\t- {k} -> {round(v, 4)}" for k, v in sample_coefficients.items()))
+
+    data_dirs = list(map(lambda x: x.path, huggingface_hub.list_repo_tree(args.dolma_hf_path, repo_type="dataset")))
```

Comment:
It probably doesn't help much, but I found any repeated calls to the hub would rate-limit me, so even caching this could be helpful (although it looks like it's just called once, so probably not worth it)

Reply:
> It probably doesn't help much, but I found any repeated calls to the hub would rate-limit me, so even caching this could be helpful (although it looks like it's just called once, so probably not worth it)

Thanks -  it's only called once and haven't faced issues with this so for now will keep it as is